File: windows/MyWhisper/AppState.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using System.Windows.Threading;
using MyWhisper.Models;
using MyWhisper.Services;
using NAudio.CoreAudioApi;

namespace MyWhisper;

/// <summary>
/// Central observable state and orchestration: wires the hotkey monitor to the
/// record → transcribe → polish → type pipeline, and exposes everything the
/// setup window and tray icon bind to.
/// </summary>
public sealed class AppState : INotifyPropertyChanged, IDisposable
{
    private const int MinSampleCount = 16_000 / 4; // 0.25 s at 16 kHz
    private const double PolishTimeoutSeconds = 3.0;

    private readonly Dispatcher _dispatcher;
    private readonly Settings _settings;
    private readonly AudioRecorder _recorder = new();
    private readonly HotkeyMonitor _hotkey = new();
    private readonly ModelDownloader _downloader = new();
    private WhisperEngine? _whisper;

    public AppState(Dispatcher dispatcher)
    {
        _dispatcher = dispatcher;
        _settings = Settings.Load();
        _selectedInputDeviceId = _settings.SelectedInputDeviceId;
        _polishEnabled = _settings.PolishEnabled;
        _polishEndpoint = _settings.PolishEndpoint;
        _polishModel = _settings.PolishModel;
        _polishApiKeyPresent = CredentialStore.HasApiKey;

        _hotkey.Pressed += () => _dispatcher.InvokeAsync(BeginRecording);
        _hotkey.Released += () => _dispatcher.InvokeAsync(EndRecordingAndTranscribeAsync);
    }

    // ── Observable properties ───────────────────────────────────────────────

    private DictationState _dictation = DictationState.Idle;
    public DictationState Dictation { get => _dictation; private set => Set(ref _dictation, value); }

    private bool _modelDownloaded;
    public bool ModelDownloaded { get => _modelDownloaded; private set { if (Set(ref _modelDownloaded, value)) Raise(nameof(IsFullyConfigured)); } }

    private double _modelDownloadProgress;
    public double ModelDownloadProgress { get => _modelDownloadProgress; private set => Set(ref _modelDownloadProgress, value); }

    private bool _hotkeyRunning;
    public bool HotkeyRunning { get => _hotkeyRunning; private set { if (Set(ref _hotkeyRunning, value)) Raise(nameof(IsFullyConfigured)); } }

    public bool IsFullyConfigured => ModelDownloaded && HotkeyRunning;

    public ObservableCollection<AudioInputDevice> AvailableInputDevices { get; } = new();

    private string? _selectedInputDeviceId;
    public string? SelectedInputDeviceId
    {
        get => _selectedInputDeviceId;
        set { if (Set(ref _selectedInputDeviceId, value)) { _settings.SelectedInputDeviceId = value; _settings.Save(); } }
    }

    private bool _polishEnabled;
    public bool PolishEnabled
    {
        get => _polishEnabled;
        set { if (Set(ref _polishEnabled, value)) { _settings.PolishEnabled = value; _settings.Save(); } }
    }

    private string _polishEndpoint;
    public string PolishEndpoint
    {
        get => _polishEndpoint;
        set { if (Set(ref _polishEndpoint, value)) { _settings.PolishEndpoint = value; _settings.Save(); } }
    }

    private string _polishModel;
    public string PolishModel
    {
        get => _polishModel;
        set { if (Set(ref _polishModel, value)) { _settings.PolishModel = value; _settings.Save(); } }
    }

    private bool _polishApiKeyPresent;
    public bool PolishApiKeyPresent { get => _polishApiKeyPresent; private set => Set(ref _polishApiKeyPresent, value); }

    private string _lastRawTranscript = "";
    public string LastRawTranscript { get => _lastRawTranscript; private set => Set(ref _lastRawTranscript, value); }

    private string _lastTranscript = "";
    public string LastTranscript { get => _lastTranscript; private set => Set(ref _lastTranscript, value); }

    private int _totalTranscriptions;
    public int TotalTranscriptions { get => _totalTranscriptions; private set => Set(ref _totalTranscriptions, value); }

    // ── Lifecycle ───────────────────────────────────────────────────────────

    /// <summary>Loads persisted state, the model, and starts the hotkey monitor.</summary>
    public async Task BootstrapAsync()
    {
        RefreshInputDevices();
        ModelDownloaded = ModelDownloader.IsDownloaded;

        if (ModelDownloaded)
            await LoadWhisperAsync();

        StartHotkey();
    }

    public void RefreshInputDevices()
    {
        var devices = AudioDeviceProbe.ListInputDevices();
        AvailableInputDevices.Clear();
        foreach (var d in devices)
            AvailableInputDevices.Add(d);

        // Drop a stale selection if that device is gone.
        if (_selectedInputDeviceId is not null &&
            !devices.Any(d => d.Id == _selectedInputDeviceId))
        {
            SelectedInputDeviceId = null;
        }
    }

    public async Task DownloadModelAsync()
    {
        try
        {
            var progress = new Progress<double>(p => ModelDownloadProgress = p);
            await _downloader.DownloadAsync(progress);
            ModelDownloaded = true;
            ModelDownloadProgress = 1.0;
            await LoadWhisperAsync();
        }
        catch (Exception ex)
        {
            Log.Error($"Model download failed: {ex.Message}");
            Dictation = DictationState.Error($"Model download failed: {ex.Message}");
        }
    }

    public void SetPolishApiKey(string? key)
    {
        CredentialStore.SetApiKey(key);
        PolishApiKeyPresent = CredentialStore.HasApiKey;
    }

    private async Task LoadWhisperAsync()
    {
        try
        {
            string path = ModelDownloader.LocalPath;
            _whisper = await Task.Run(() => new WhisperEngine(path));
            Log.Info("Whisper model loaded");
        }
        catch (Exception ex)
        {
            Log.Error($"Failed to load Whisper model: {ex.Message}");
            Dictation = DictationState.Error($"Failed to load model: {ex.Message}");
        }
    }

    private void StartHotkey()
    {
        try
        {
            _hotkey.Start();
            HotkeyRunning = true;
            Log.Info("Hotkey monitor started (Right Alt)");
        }
        catch (Exception ex)
        {
            Log.Error($"Hotkey monitor failed: {ex.Message}");
            Dictation = DictationState.Error($"Hotkey monitor failed: {ex.Message}");
        }
    }

    // ── Dictation pipeline ──────────────────────────────────────────────────

    private void BeginRecording()
    {
        if (!ModelDownloaded || _whisper is null)
        {
            Log.Warn("BeginRecording skipped: model not ready");
            return;
        }
        if (Dictation.Phase is not (DictationPhase.Idle or DictationPhase.NoSpeech))
        {
            Log.Info($"BeginRecording skipped: phase={Dictation.Phase}");
            return;
        }

        try
        {
            MMDevice? device = _selectedInputDeviceId is not null
                ? AudioDeviceProbe.DeviceById(_selectedInputDeviceId)
                : null;
            device ??= AudioDeviceProbe.DefaultInputDevice();
            if (device is null)
            {
                Dictation = DictationState.Error("No microphone available");
                return;
            }

            _recorder.Start(device);
            Dictation = DictationState.Recording;
            Log.Info($"Recording started on '{device.FriendlyName}'");
        }
        catch (Exception ex)
        {
            Log.Error($"Recording failed to start: {ex.Message}");
            Dictation = DictationState.Error($"Recording failed: {ex.Message}");
        }
    }

    private async Task EndRecordingAndTranscribeAsync()
    {
        if (Dictation.Phase != DictationPhase.Recording)
        {
            Log.Info($"EndRecording skipped: phase={Dictation.Phase}");
            return;
        }

        float[] samples;
        try
        {
            samples = await _recorder.StopAsync();
        }
        catch (Exception ex)
        {
            Log.Error($"StopAsync failed: {ex.Message}");
            Dictation = DictationState.Error($"Recording failed: {ex.Message}");
            return;
        }

        var (rms, peak) = Amplitude(samples);
        Log.Info($"Recording stopped: {samples.Length} samples " +
                 $"({samples.Length / 16000.0:F2}s) rms={rms:F4} peak={peak:F4} " +
                 $"device='{_recorder.LastDeviceName}'");

        if (samples.Length < MinSampleCount)
        {
            Log.Info("Samples too short, ignoring");
            await FlashNoSpeechAsync();
            return;
        }
        if (peak < 0.005f)
            Log.Warn($"Audio near-silent (peak={peak:F4}); mic may be muted or wrong device selected");

        Dictation = DictationState.Transcribing;

        string raw;
        try
        {
            var whisper = _whisper!;
            raw = await Task.Run(() => whisper.TranscribeAsync(samples));
        }
        catch (Exception ex)
        {
            Log.Error($"Transcription failed: {ex.Message}");
            Dictation = DictationState.Error($"Transcription failed: {ex.Message}");
            return;
        }
        Log.Info($"Whisper raw: \"{raw}\"");

        string cleaned = Clean(raw);
        if (cleaned.Length == 0)
        {
            Log.Info("Transcript empty after cleaning, no paste");
            await FlashNoSpeechAsync();
            return;
        }

        LastRawTranscript = cleaned;
        string polished = await PolishIfEnabledAsync(cleaned);

        Dictation = DictationState.Typing;
        LastTranscript = polished;
        TotalTranscriptions++;
        Log.Info($"Typing: \"{polished}\"");

        try
        {
            TextInjector.Type(polished);
        }
        catch (Exception ex)
        {
            Log.Error($"Text injection failed: {ex.Message}");
        }

        await Task.Delay(250);
        Dictation = DictationState.Idle;
    }

    private async Task<string> PolishIfEnabledAsync(string raw)
    {
        if (!PolishEnabled || !PolishApiKeyPresent)
            return raw;

        string? key = CredentialStore.GetApiKey();
        if (string.IsNullOrEmpty(key) || !Uri.TryCreate(PolishEndpoint, UriKind.Absolute, out _))
            return raw;

        try
        {
            var config = new PolisherConfig(
                PolishEndpoint, PolishModel, key, TimeSpan.FromSeconds(PolishTimeoutSeconds));
            string polished = await Polisher.PolishAsync(raw, config);
            return string.IsNullOrWhiteSpace(polished) ? raw : polished;
        }
        catch (Exception ex)
        {
            Log.Error($"Polish failed, falling back to raw: {ex.Message}");
            return raw;
        }
    }

    private async Task FlashNoSpeechAsync()
    {
        Dictation = DictationState.NoSpeech;
        await Task.Delay(800);
        Dictation = DictationState.Idle;
    }

    private static string Clean(string raw)
    {
        string text = raw.Trim();
        string[] junk = { "[BLANK_AUDIO]", "[silence]", "(silence)", "[ Silence ]", "[no audio]" };
        foreach (string j in junk)
            text = text.Replace(j, "", StringComparison.OrdinalIgnoreCase);
        return text.Trim();
    }

    private static (float Rms, float Peak) Amplitude(float[] samples)
    {
        if (samples.Length == 0)
            return (0f, 0f);
        float sumSq = 0f, peak = 0f;
        foreach (float s in samples)
        {
            sumSq += s * s;
            float a = MathF.Abs(s);
            if (a > peak) peak = a;
        }
        return (MathF.Sqrt(sumSq / samples.Length), peak);
    }

    // ── INotifyPropertyChanged ──────────────────────────────────────────────

    public event PropertyChangedEventHandler? PropertyChanged;

    private bool Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
            return false;
        field = value;
        Raise(name);
        return true;
    }

    private void Raise(string? name)
    {
        if (_dispatcher.CheckAccess())
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        else
            _dispatcher.InvokeAsync(() => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name)));
    }

    public void Dispose()
    {
        _hotkey.Dispose();
        _recorder.Dispose();
        _whisper?.Dispose();
    }
}

File: macos/MyWhisper/AppState.swift
import Foundation
import SwiftUI
import AVFoundation
import OSLog

private let log = Logger(subsystem: "com.mywhisper.app", category: "AppState")

enum DictationState: Equatable {
    case idle
    case recording
    case transcribing
    case typing
    case noSpeech
    case error(String)
}

@MainActor
final class AppState: ObservableObject {
    static let shared = AppState()
    private static let selectedDeviceUIDKey = "selectedInputDeviceUID"
    private static let selectedLanguageKey  = "selectedLanguage"
    private static let selectedModelIDKey   = "selectedModelID"
    private static let polishEnabledKey      = "polishEnabled"
    private static let polishEndpointKey     = "polishEndpoint"
    private static let polishModelKey        = "polishModel"

    @Published var dictation: DictationState = .idle
    @Published var micPermission: PermissionState = .unknown
    @Published var accessibilityGranted: Bool = false
    @Published var modelDownloaded: Bool = false
    @Published var modelDownloadProgress: Double = 0
    @Published var lastTranscript: String = ""
    @Published var lastRawTranscript: String = ""
    @Published var totalTranscriptions: Int = 0
    @Published var availableInputDevices: [AudioInputDevice] = []
    @Published var selectedInputDeviceUID: String? {
        didSet { UserDefaults.standard.set(selectedInputDeviceUID, forKey: Self.selectedDeviceUIDKey) }
    }
    @Published var selectedLanguage: String = "auto" {
        didSet { UserDefaults.standard.set(selectedLanguage, forKey: Self.selectedLanguageKey) }
    }
    @Published var selectedModelID: String = "base" {
        didSet {
            UserDefaults.standard.set(selectedModelID, forKey: Self.selectedModelIDKey)
            Task { @MainActor in await self.onSelectedModelChanged() }
        }
    }

    var selectedModel: WhisperModel { WhisperModel.model(id: selectedModelID) }

    /// Name of the system default input device, or nil if unavailable.
    var defaultInputDeviceName: String? {
        AudioDeviceProbe.defaultInputDeviceName()
    }

    /// Human-readable name of the active input device.
    var activeInputDeviceName: String {
        if let uid = selectedInputDeviceUID,
           let device = availableInputDevices.first(where: { $0.uid == uid }) {
            return device.name
        }
        return defaultInputDeviceName ?? "Unknown"
    }

    @Published var polishEnabled: Bool = false {
        didSet { UserDefaults.standard.set(polishEnabled, forKey: Self.polishEnabledKey) }
    }
    @Published var polishEndpoint: String = "https://ollama.com/v1/chat/completions" {
        didSet { UserDefaults.standard.set(polishEndpoint, forKey: Self.polishEndpointKey) }
    }
    @Published var polishModel: String = "gpt-oss:20b" {
        didSet { UserDefaults.standard.set(polishModel, forKey: Self.polishModelKey) }
    }
    @Published var polishAPIKeyPresent: Bool = false

    private let recorder = AudioRecorder()
    private var whisper: WhisperEngine?
    private let hotkey = HotkeyMonitor()
    private let downloader = ModelDownloader()
    private var permissionPollTimer: Timer?

    private init() {
        let defaults = UserDefaults.standard
        self.selectedInputDeviceUID = defaults.string(forKey: Self.selectedDeviceUIDKey)
        if let lang = defaults.string(forKey: Self.selectedLanguageKey), !lang.isEmpty {
            self.selectedLanguage = lang
        }
        if let mid = defaults.string(forKey: Self.selectedModelIDKey), !mid.isEmpty {
            self.selectedModelID = mid
        }
        self.polishEnabled  = defaults.bool(forKey: Self.polishEnabledKey)
        if let ep = defaults.string(forKey: Self.polishEndpointKey), !ep.isEmpty {
            self.polishEndpoint = ep
        }
        if let m = defaults.string(forKey: Self.polishModelKey), !m.isEmpty {
            self.polishModel = m
        }
        self.polishAPIKeyPresent = KeychainStore.apiKey() != nil
    }

    func setPolishAPIKey(_ key: String?) {
        KeychainStore.setAPIKey(key)
        polishAPIKeyPresent = KeychainStore.apiKey() != nil
    }

    private func polishIfEnabled(_ raw: String) async -> String {
        guard polishEnabled, polishAPIKeyPresent,
              let url = URL(string: polishEndpoint),
              let key = KeychainStore.apiKey() else { return raw }

        let config = PolisherConfig(endpoint: url, model: polishModel, apiKey: key, timeout: 3.0)
        do {
            let polished = try await Polisher.polish(raw, config: config)
            return polished.isEmpty ? raw : polished
        } catch {
            log.error("polish failed, falling back to raw: \(error.localizedDescription, privacy: .public)")
            return raw
        }
    }

    var isFullyConfigured: Bool {
        micPermission == .granted && accessibilityGranted && modelDownloaded
    }

    var menuBarSymbol: String {
        switch dictation {
        case .idle: return isFullyConfigured ? "mic" : "mic.slash"
        case .recording: return "mic.fill"
        case .transcribing: return "waveform"
        case .typing: return "keyboard"
        case .noSpeech: return "mic.badge.xmark"
        case .error: return "exclamationmark.triangle"
        }
    }

    func bootstrap() async {
        refreshMicPermission()
        refreshAccessibility()
        refreshModelStatus()
        refreshInputDevices()
        startPermissionPolling()
        if modelDownloaded {
            await loadWhisper()
        }
        if isFullyConfigured {
            startHotkey()
        }
    }

    func refreshInputDevices() {
        availableInputDevices = AudioDeviceProbe.listInputDevices()
        // Drop the selection if the device went away.
        if let uid = selectedInputDeviceUID, !availableInputDevices.contains(where: { $0.uid == uid }) {
            selectedInputDeviceUID = nil
        }
    }

    func refreshMicPermission() {
        switch AVCaptureDevice.authorizationStatus(for: .audio) {
        case .authorized: micPermission = .granted
        case .denied, .restricted: micPermission = .denied
        case .notDetermined: micPermission = .unknown
        @unknown default: micPermission = .unknown
        }
    }

    func requestMicPermission() async {
        let granted = await AVCaptureDevice.requestAccess(for: .audio)
        micPermission = granted ? .granted : .denied
        if granted {
            refreshInputDevices()
        }
        await reconcile()
        // The system permission dialog steals focus; bring the setup window back.
        NSApp.activate(ignoringOtherApps: true)
        NotificationCenter.default.post(name: AppDelegate.openSetupNotification, object: nil)
    }

    func refreshAccessibility() {
        accessibilityGranted = AXIsProcessTrusted()
    }

    func promptAccessibility() {
        let options: NSDictionary = [
            kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: true
        ]
        accessibilityGranted = AXIsProcessTrustedWithOptions(options)
    }

    func refreshModelStatus() {
        modelDownloaded = ModelDownloader.isDownloaded(selectedModel)
    }

    func downloadModel() async {
        let target = selectedModel
        modelDownloadProgress = 0
        do {
            try await downloader.download(model: target) { [weak self] progress in
                Task { @MainActor in self?.modelDownloadProgress = progress }
            }
            // Only flip downloaded=true if the user is still on the same model.
            if selectedModelID == target.id {
                modelDownloaded = true
                modelDownloadProgress = 1.0
                await loadWhisper()
                await reconcile()
            }
        } catch {
            dictation = .error("Model download failed: \(error.localizedDescription)")
        }
    }

    private func loadWhisper() async {
        let path = ModelDownloader.localPath(for: selectedModel).path
        do {
            whisper = try WhisperEngine(modelPath: path)
            log.info("loaded whisper model: \(self.selectedModelID, privacy: .public)")
        } catch {
            dictation = .error("Failed to load model: \(error.localizedDescription)")
        }
    }

    /// Called whenever the user picks a different model in the picker.
    /// If the new model is already downloaded, swap the whisper engine; otherwise
    /// surface the Download button by clearing the downloaded flag.
    private func onSelectedModelChanged() async {
        whisper = nil
        modelDownloadProgress = 0
        refreshModelStatus()
        if modelDownloaded {
            await loadWhisper()
        }
        await reconcile()
    }

    private func reconcile() async {
        if isFullyConfigured {
            startHotkey()
        }
    }

    private func startHotkey() {
        hotkey.onPress = { [weak self] in
            Task { @MainActor in await self?.beginRecording() }
        }
        hotkey.onRelease = { [weak self] in
            Task { @MainActor in await self?.endRecordingAndTranscribe() }
        }
        do {
            try hotkey.start()
        } catch {
            dictation = .error("Hotkey monitor failed: \(error.localizedDescription)")
        }
    }

    private func startPermissionPolling() {
        permissionPollTimer?.invalidate()
        permissionPollTimer = Timer.scheduledTimer(withTimeInterval: 1.5, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self else { return }
                let wasGranted = self.accessibilityGranted
                self.refreshAccessibility()
                if !wasGranted && self.accessibilityGranted {
                    await self.reconcile()
                }
            }
        }
    }

    private func beginRecording() async {
        guard isFullyConfigured else {
            log.warning("beginRecording: not fully configured (mic=\(String(describing: self.micPermission), privacy: .public), a11y=\(self.accessibilityGranted, privacy: .public), model=\(self.modelDownloaded, privacy: .public))")
            return
        }
        guard whisper != nil else {
            log.error("beginRecording: whisper engine not loaded")
            dictation = .error("Whisper model not loaded")
            return
        }
        guard dictation == .idle || dictation == .noSpeech else {
            log.debug("beginRecording: skipped, state=\(String(describing: self.dictation), privacy: .public)")
            return
        }
        do {
            let deviceID = selectedInputDeviceUID.flatMap(AudioDeviceProbe.deviceID(forUID:))
            try recorder.start(deviceID: deviceID)
            dictation = .recording
            log.info("recording started")
        } catch {
            log.error("recording failed to start: \(error.localizedDescription, privacy: .public)")
            dictation = .error("Recording failed: \(error.localizedDescription)")
        }
    }

    private func endRecordingAndTranscribe() async {
        guard dictation == .recording else {
            log.debug("endRecording: not recording (state=\(String(describing: self.dictation), privacy: .public))")
            return
        }
        let samples = recorder.stop()
        let duration = Double(samples.count) / 16_000.0
        let (rms, peak) = Self.amplitude(samples)
        log.info("recording stopped: \(samples.count, privacy: .public) samples (\(String(format: "%.2f", duration), privacy: .public)s) rms=\(String(format: "%.4f", rms), privacy: .public) peak=\(String(format: "%.4f", peak), privacy: .public)")

        guard samples.count > 16_000 / 4 else {
            log.notice("samples too short, ignoring")
            dictation = .noSpeech
            try? await Task.sleep(nanoseconds: 800_000_000)
            dictation = .idle
            return
        }
        if peak < 0.005 {
            log.notice("audio is near silent (peak=\(String(format: "%.4f", peak), privacy: .public)). Mic not capturing.")
        }
        dictation = .transcribing
        let engine = whisper
        let language = selectedLanguage
        let raw: String = await Task.detached(priority: .userInitiated) { [samples, language] in
            guard let engine else { return "" }
            do {
                return try engine.transcribe(samples: samples, language: language)
            } catch {
                log.error("whisper transcribe threw: \(error.localizedDescription, privacy: .public)")
                return ""
            }
        }.value
        log.info("whisper raw (\(language, privacy: .public)): \"\(raw, privacy: .public)\"")

        let cleaned = Self.clean(raw)
        guard !cleaned.isEmpty else {
            log.notice("transcript empty after cleaning, no paste")
            dictation = .noSpeech
            try? await Task.sleep(nanoseconds: 800_000_000)
            dictation = .idle
            return
        }
        lastRawTranscript = cleaned

        dictation = .transcribing
        let polished = await polishIfEnabled(cleaned)

        dictation = .typing
        lastTranscript = polished
        totalTranscriptions += 1
        log.info("pasting: \"\(polished, privacy: .public)\"")
        TextInjector.paste(polished)
        try? await Task.sleep(nanoseconds: 250_000_000)
        dictation = .idle
    }

    private static func amplitude(_ samples: [Float]) -> (rms: Float, peak: Float) {
        guard !samples.isEmpty else { return (0, 0) }
        var sumSq: Float = 0
        var peak: Float = 0
        for s in samples {
            sumSq += s * s
            let a = abs(s)
            if a > peak { peak = a }
        }
        return (sqrt(sumSq / Float(samples.count)), peak)
    }

    private static func clean(_ raw: String) -> String {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        // Whisper sometimes emits "[BLANK_AUDIO]" or "(silence)" for empty input.
        let junk = ["[BLANK_AUDIO]", "[silence]", "(silence)", "[ Silence ]", "[no audio]"]
        for j in junk where text.localizedCaseInsensitiveContains(j) {
            text = text.replacingOccurrences(of: j, with: "", options: .caseInsensitive)
        }
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

enum PermissionState {
    case unknown
    case granted
    case denied
}
